src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,505 @@
+                {
+                    Debug.Assert(boundNode != null);
+
+                    Parent = parent;
+                    BoundNode = boundNode;
+                    ContainingClosure = containingClosure;
+                }
+
+                public void Free()
+                {
+                    foreach (var scope in NestedScopes)
+                    {
+                        scope.Free();
+                    }
+                    NestedScopes.Free();
+
+                    foreach (var closure in Closures)
+                    {
+                        closure.Free();
+                    }
+                    Closures.Free();
+                    DeclaredVariables.Free();
+                }
+
+                public override string ToString() => BoundNode.Syntax.GetText().ToString();
+            }
+
+            /// <summary>
+            /// The Closure type represents a lambda or local function and stores
+            /// information related to that closure. After initially building the
+            /// <see cref="Scope"/> tree the only information available is
+            /// <see cref="OriginalMethodSymbol"/> and <see cref="CapturedVariables"/>.
+            /// Subsequent passes are responsible for translating captured
+            /// variables into captured environments and for calculating
+            /// the rewritten signature of the method.
+            /// </summary>
+            private sealed class Closure
+            {
+                /// <summary>
+                /// The method symbol for the original lambda or local function.
+                /// </summary>
+                public MethodSymbol OriginalMethodSymbol { get; }
+
+                public PooledHashSet<Symbol> CapturedVariables { get; } = PooledHashSet<Symbol>.GetInstance();
+
+                public Closure(MethodSymbol symbol)
+                {
+                    Debug.Assert(symbol != null);
+                    OriginalMethodSymbol = symbol;
+                }
+
+                public void Free()
+                {
+                    CapturedVariables.Free();
+                }
+            }
+
+            /// <summary>
+            /// Optimizes local functions such that if a local function only references other local functions
+            /// that capture no variables, we don't need to create capture environments for any of them.
+            /// </summary>
+            private void RemoveUnneededReferences()
+            {
+                var methodGraph = new MultiDictionary<MethodSymbol, MethodSymbol>();
+                var capturesThis = new HashSet<MethodSymbol>();
+                var capturesVariable = new HashSet<MethodSymbol>();
+                var visitStack = new Stack<MethodSymbol>();
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    foreach (var capture in closure.CapturedVariables)
```

Comment:
OK, here's my argument for why this doesn't change the behavior:

This loop can have three possible side effects.

1. The capture is added to `methodGraph`. This is itself an unordered data structure and it isn't read from inside the loop, so the results are also unordered.
2. The capture is added to `capturesVariables`, which is also unordered and is only read from to determine `visitStack`.
3. The capture is added to `capturesThis`. The same argument as (2).

This leaves `visitStack`, which is order dependent, so we have to prove the contents are not used in a way that is order-sensitive.

First, the `while` loop. Here we calculate `setToAddTo` based on `capturesVariable.Contains`, which we have already established is not order dependent. Then, we visit `methodGraph[current]` in no particular order, the contents of which are independent of the earlier visitation order, as we already showed. We then add each `capturesCurrent` to `setToAddTo`, which is always unordered. If `setToAddTo` doesn't contain the result we add it to the stack. As we already established, the initial contents of `setToAddTo` on entering the `while` loop are invariant. This leaves the question of intermediate runs of the `foreach` loop. Since the `visitStack` is ordering dependent, the contents of `setToAddTo` could change depending on which is visited first. However, since anything *not* already visited in the stack is merely added to the stack to be visited later, and all side effects are added to order-independent, item-exclusive sets, the presence of variables should be unaffected by whether or not they're visited first or later. Essentially, even if the stack is visited out of order, all of the elements in the stack will eventually be visited and the resulting sets don't care about addition order.

Finally, the last `foreach` loop doesn't use the captured variables list directly and only depends on the unordered `capturesVariable` and `capturesThis` lists, which we've established have the same contents as before.